### PR TITLE
iox-#2177 Update SPSC SoFi

### LINKED ISSUE #1695: Decipher weird comment in SoFi
## Required information

In the sofi implementation `sofi.inl` is a weird todo which states:
```
 /// @todo read before write since the writer increments the aba counter!!!
 /// @todo write doc with example!!!
```

Unravel hidden meaning and write doc with example has hinted in to todo ... Or do nothing if it is not relevant anymore.

### LINKED ISSUE #2160: Bug in lock-free sofi
## Required information

Assume the following scenario: A new subscriber subscribes to a service with a history size of `N`. The subscriber's buffer has a capacity of `N` as well, and the history is full.

1. RouDi calls `SoFi::push` until sofi is full and the next push call would cause an overflow. The memory is correctly synced with `m_write_pos.store(std::memory_order_release)`
2. The subscriber does not consume the history or anything in the buffer, sofi remains full.
3. The publisher in another process and therefore also in another thread publishes a new sample. Since sofi is in overflow, the oldest value must copied out and returned to the user. But there is no corresponding `m_write_pos.load(std::memory_order_acquire)` call to sync the `m_data` memory from the roudi thread to the publisher thread. The publisher may return an unsynced garbage value.

See: https://github.com/eclipse-iceoryx/iceoryx/blob/2c612f5a59e23f221fbbd55235c58dcc17dd3830/iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl#L143

So, whenever the `push` method in sofi is called from multiple threads non-concurrently, those threads must sync `m_data` for the overflow case.

### LINKED ISSUE #2167: Update SPSCFiFo
## Brief feature description

The goal of this issue is to improve the documentation of the FiFo as well as to fix potential wrong memory orders

## Detailed information

- better understanding of the memory orders
- fix potential data races
- relax memory orders when possible

### PR DESCRIPTION
**Pull request by albtam**
- add comments to clarify concurrent behavior
- fix memory orders

## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

I updated some memory orders and updated the comments to make sure future contributors understand the changes

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes #2167 
- Closes #2160 
- Closes #1695

### CODE REVIEW

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:144]: From https://github.com/eclipse-iceoryx/iceoryx/pull/2178#discussion_r1476417107


> I have to think about this change a bit more. We might still need to synchronize with a corresponding m_readPos release operation.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:154]: From https://github.com/eclipse-iceoryx/iceoryx/pull/2178#discussion_r1476461445

> Please revert all the changes from this method. I would bet the crate of beer @budrus still owes me that test_stress_spsc_sofi.cpp will fail when run on a CPU with a weakly ordered CPU like on an armv8.

**Comment by albtam** [comment]
From https://github.com/eclipse-iceoryx/iceoryx/pull/2178#issuecomment-1929662893

@elBoberido: 

> I could make your example pass miri by making both compare and swap loops use either compare_exchange or compare_exchange_weak.
> But when increasing the number of pushes a lot it still detects (false positive?) races. The tests also need to be done in a different way, e.g. the cancelation points of the loop become more complex due to the overflow behavior.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:177]: Please change this to a similar phrase like in the SpscFifo. This is semantically not correct since the consumer does not yet know about the stored value and also can not pop it. As far as it concerns the consumer, it doesn't matter when the value is stored in the buffer. Theoretically the producer could be suspended for hours before the write position is advanced, e.g. by setting a break point at the next line.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:186]: This needs to be reverted since it acts as sync point to the CAS loop in the pop method.

**Review by elBoberido — COMMENTED** [review]
Sorry for letting you wait so long for the review. I needed a fresh mind to review lock-free code :)

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:110]: Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:92]: Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:159]: Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:158]: Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:51]: After a second thought, I am not sure the diagram clarifies the comment. I think it's enough to state "It's an expected behavior that when push/pop are called concurrently and SpscSoFi is full, as many elements as specified with 'CapacityValue' can be removed"

I removed the diagram

**Comment by elBoberido** [comment]
@albtam just to be sure that you are not waiting on feedback. I'll wait until the CI (at least the github actions) are green.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:116]: Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:204]: Reverted

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:131]: Reverted

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:127]: Done

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:175]: Done

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: @elBoberido What do you recommend to do then?

**Comment by albtam** [comment]
> @albtam just to be sure that you are not waiting on feedback. I'll wait until the CI (at least the github actions) are green.

@elBoberido Back to you

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Exactly what you did, change the memory order back to `memory_order_acq_rel` and `memory_order_acquire` :)

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:117]: Can you move the explanation to
```cpp
        if (currentReadPosition == m_writePosition.load(std::memory_order_acquire))
        {
            nextReadPosition = currentReadPosition;
            popWasSuccessful = false;
        }
```
below. It is not related to the memory order but the reason on cannot just call `return false;` but needs to continue with the loop.

The `acquire` just corresponds to the `m_readPosition.compare_exchange_strong` store operation in push.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:189]: This corresponds to the `m_readPosition.compare_exchange_weak` store operation in pop.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:143]: No, `acquire` is required in order to not reorder the memcpy after the CAS operation. With `relaxed` memory order it would be perfectly fine for the CPU/compiler to perform the CAS Operation and do the memcpy only after the CAS operation was successful in order to not do useless memcpy's

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:40]: removed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:92]: changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.hpp:50]: changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:189]: Yes, what I wanted to say is that the memory synchronization is not required. But to avoid the reordering, we need a corresponding store/release. So in the end, there will be a memory synchronization but this is only a side effect. Do you agree?

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:143]: You're right. Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:117]: I don't understand. You're saying that without any operation reordering and by returning false, we can have the following scenario:

```
0. Initial situation (the queue is full)
|----|--B--|--C--|
^    ^
w=3 r=1
1. The consumer thread loads m_writePosition => 3
|----|--B--|--C--|
^     ^
w=3  r=1
2. The producer thread pushes two times
|--D--|--E--|-----|
^           ^
r=3        w=5
3. The consumer thread loads m_readPosition == m_writePosition so the pop method returns false
=> Whereas the queue was full, pop returned false giving the impression that the queue if empty
```

**Comment by albtam** [comment]
@elBoberido Thanks. I addressed most of your comments. There are still some points to clarify

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Chnaged

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: You are right. I confused it with a previous version of the patch where the load of `m_readPosition` was relaxed. With `acquire` this cannot happen. I think it should now even be possible to have an early return. Unfortunately we do not have the full history of the SpscSofi but I think we once had a slightly different algorithm and kept `popWasSuccessful` after the refactoring.

The comment should be moved back where it was and an early return can be introduced.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Yes, I agree. What do you think about talking about `happens before` relationships instead of `memory synchronization` and `reordering`? This is a more generic term and from the context it becomes clear if the happens before relationship is required for memory or for logic.

In this specific case the comment could become something like `We need to establish a happens before relationship with 'm_writePosition.load(...)'`

**Comment by elBoberido** [comment]
@albtam you can undo some of my suggestions :sweat_smile:

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: @elBoberido Sorry, can you remind me what's the happens-before relationship we want to enforce here? Since it's  a load/acquire, it doesn't influence the order of operations above (so it can't be `m_writePosition.load(...)`).

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Hehe, now you get some warnings because `nextReadPosition` is set but not used anymore and the clang-tidy also does not like to have an else branch after an early return :)

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [inline iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:147]: It's not only the corruption of `valueOut` but we also need to prevent to reorder the memcpy after updating `m_readPosition`. With `relaxed` it would be perfectly fine for the compiler and CPU to successfully finish the while loop and then do the memcpy only once afterwards.

**Review by elBoberido — COMMENTED** [review]
This somehow becomes a never ending story but with the revert of the `relaxed` ordering in the places where `acq_rel` was required, we might be able to relax some of the `acquire` orderings. Unfortunately this is something which is hard to see when there are too many wrong orderings but once the ordering is at least as strong as required, we can identify the locations which definitely need the strong ordering and can try to relax some orderings which act as sync points when we identify new sync points or change the algorithm a bit.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:147]: Correct

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Changed

**Comment by albtam** [comment]
I pushed some of your proposals. There is still something unclear. Lock-free is hard :P

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (elBoberido) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: I mean we can make `uint64_t currentReadPosition = m_readPosition.load(std::memory_order_acquire);` in `pop` also `relaxed` but then the early return needs to be removed again and the previous logic with `popWasSuccessful` needs to be re-introduced.

Overall, this would lead to only one `acquire` and one `release` for `m_writePosition` and two `acq_rel`/`acquire` for `m_readPosition` which simplifies the relation of the sync points.

**Review by elBoberido — COMMENTED** [review]
Yes, lock-free is hard :sweat_smile: 

Btw, can you please adjust your commit messages with the issue number.

Before this is merged, it would be good to run `test_stress_spsc_sofi`. Ideally each of the three test for a few hours. The stress time can only be adjusted in the source code.

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl]: Ok, changed

File: iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl
Comment (albtam) [reply iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl:209]: Changed

**Comment by albtam** [comment]
@elBoberido I implemented your last recommendations.

**Comment by elBoberido** [comment]
@albtam did you have the chance to run the stress tests on an aarch64 CPU?

**Comment by albtam** [comment]
@elBoberido Sorry for the delay. I ran the stress tests (on x64 CPU). This test failed SpscSofiStress.SimultaneouslyPushAndPopOnEmptySoFi (both on master and my branch).

I don't have an aarch64 CPU, would it be possible for you to run the tests?

**Comment by orecham** [comment]
> @elBoberido Sorry for the delay. I ran the stress tests (on x64 CPU). This test failed SpscSofiStress.SimultaneouslyPushAndPopOnEmptySoFi (both on master and my branch).
> 
> I don't have an aarch64 CPU, would it be possible for you to run the tests?

@albtam I can run it for you on my mac

**Comment by elBoberido** [comment]
If it runs without error on the mac, then this can be merged. But before it needs to be rebased. There is now an abstraction for the atomics which leads to a merge conflict. It should be easily solved, though.

@albtam can you post the output of the stress tests. I've run main and your branch and didn't encounter an error.

**Comment by albtam** [comment]
@elBoberido 

> I've run main and your branch and didn't encounter an error.

I reran them and everything's fine. 

I fixed the conflicts. Back to you

**Review by elBoberido — APPROVED** [review]

**Comment by elBoberido** [comment]
@albtam thanks for your contribution :)

**Pull request merged**
